File: inference/src/armor_inference/hardware.py
"""Best-effort host hardware inventory: CPU, RAM, GPU.

Standard library only, plus whatever ML stack this build already has
(`onnxruntime`, if the `onnx`/`cuda` extra is installed) — matching the base
image's zero-new-dependency ethos (see `main.py`'s module docstring). GPU
detection shells out to `nvidia-smi` when present; a host with no GPU (or no
`nvidia-smi` on PATH) just reports an empty `gpus` list rather than failing
the request.
"""

from __future__ import annotations

import os
import platform
import shutil
import subprocess
from typing import List, Optional, Tuple

from pydantic import BaseModel


class CpuInfo(BaseModel):
    model: Optional[str] = None
    architecture: str
    logical_cores: Optional[int] = None
    physical_cores: Optional[int] = None


class MemoryInfo(BaseModel):
    total_bytes: Optional[int] = None


class GpuInfo(BaseModel):
    name: str
    memory_total_mb: Optional[int] = None
    driver_version: Optional[str] = None


class HardwareInfo(BaseModel):
    cpu: CpuInfo
    memory: MemoryInfo
    gpus: List[GpuInfo]
    os: str
    python_version: str
    onnxruntime_version: Optional[str] = None
    onnxruntime_providers: List[str] = []


def _linux_cpuinfo() -> Tuple[Optional[str], Optional[int]]:
    """`(model name, physical core count)` from `/proc/cpuinfo`, best-effort.

    Physical cores is the count of distinct (physical id, core id) pairs
    across logical processors — plain `os.cpu_count()` counts hyperthreads
    too, which overstates cores on any SMT-enabled host.
    """
    try:
        text = open("/proc/cpuinfo").read()
    except OSError:
        return None, None

    model: Optional[str] = None
    pairs = set()
    phys_id = core_id = None
    for line in text.splitlines():
        if not line.strip():
            if phys_id is not None and core_id is not None:
                pairs.add((phys_id, core_id))
            phys_id = core_id = None
            continue
        key, sep, value = line.partition(":")
        if not sep:
            continue
        key = key.strip().lower()
        value = value.strip()
        if key == "model name" and model is None:
            model = value
        elif key == "physical id":
            phys_id = value
        elif key == "core id":
            core_id = value
    if phys_id is not None and core_id is not None:
        pairs.add((phys_id, core_id))
    return model, (len(pairs) or None)


def _linux_mem_total_bytes() -> Optional[int]:
    try:
        with open("/proc/meminfo") as f:
            for line in f:
                if line.startswith("MemTotal:"):
                    return int(line.split()[1]) * 1024
    except (OSError, ValueError, IndexError):
        pass
    return None


def _nvidia_gpus() -> List[GpuInfo]:
    """Queries `nvidia-smi` if it's on PATH. Absent binary, no GPU, or a
    timed-out/failed call all just mean "no GPUs to report" — this must
    never turn a CPU-only host into a 500."""
    exe = shutil.which("nvidia-smi")
    if not exe:
        return []
    try:
        out = subprocess.run(
            [
                exe,
                "--query-gpu=name,memory.total,driver_version",
                "--format=csv,noheader,nounits",
            ],
            capture_output=True,
            text=True,
            timeout=5,
            check=True,
        )
    except (OSError, subprocess.SubprocessError):
        return []

    gpus = []
    for line in out.stdout.strip().splitlines():
        parts = [p.strip() for p in line.split(",")]
        if len(parts) < 2 or not parts[0]:
            continue
        try:
            memory_mb: Optional[int] = int(float(parts[1]))
        except ValueError:
            memory_mb = None
        gpus.append(
            GpuInfo(
                name=parts[0],
                memory_total_mb=memory_mb,
                driver_version=parts[2] if len(parts) > 2 and parts[2] else None,
            )
        )
    return gpus


def _onnxruntime_info() -> Tuple[Optional[str], List[str]]:
    try:
        import onnxruntime as ort
    except ImportError:
        return None, []
    return ort.__version__, list(ort.get_available_providers())


def get_hardware_info() -> HardwareInfo:
    model, physical_cores = (None, None)
    if platform.system() == "Linux":
        model, physical_cores = _linux_cpuinfo()
    if model is None:
        model = platform.processor() or None

    mem_total = _linux_mem_total_bytes() if platform.system() == "Linux" else None
    ort_version, ort_providers = _onnxruntime_info()

    return HardwareInfo(
        cpu=CpuInfo(
            model=model,
            architecture=platform.machine(),
            logical_cores=os.cpu_count(),
            physical_cores=physical_cores,
        ),
        memory=MemoryInfo(total_bytes=mem_total),
        gpus=_nvidia_gpus(),
        os=f"{platform.system()} {platform.release()}",
        python_version=platform.python_version(),
        onnxruntime_version=ort_version,
        onnxruntime_providers=ort_providers,
    )

File: crates/api/src/hardware.rs
//! Best-effort local host hardware inventory (CPU, RAM, GPU) for
//! `GET /api/v1/hardware`.
//!
//! Core and the inference sidecar are commonly deployed on different
//! hardware (a CPU-only host running `armor-api` next to a GPU box running
//! `armor-inference`, or vice versa), so this reports *this process's* host
//! only — a sibling to `armor_inference.hardware.get_hardware_info()`, which
//! does the same for the sidecar's own host. `control_plane.rs` combines
//! both into one response so the UI can show them side by side rather than
//! one number implying they share a machine.
//!
//! Standard library plus a `nvidia-smi` shell-out only — no new crate for a
//! nice-to-have, matching this service's dependency posture elsewhere.

use std::{collections::HashSet, fs, process::Command};

use serde::Serialize;

#[derive(Serialize)]
pub struct CpuInfo {
    pub model: Option<String>,
    pub architecture: String,
    pub logical_cores: Option<usize>,
    pub physical_cores: Option<usize>,
}

#[derive(Serialize)]
pub struct MemoryInfo {
    pub total_bytes: Option<u64>,
}

#[derive(Serialize)]
pub struct GpuInfo {
    pub name: String,
    pub memory_total_mb: Option<u64>,
    pub driver_version: Option<String>,
}

#[derive(Serialize)]
pub struct HardwareInfo {
    pub cpu: CpuInfo,
    pub memory: MemoryInfo,
    pub gpus: Vec<GpuInfo>,
    pub os: String,
}

/// `(model name, physical core count)` from `/proc/cpuinfo`. Physical cores
/// is the count of distinct (physical id, core id) pairs across logical
/// processors — `available_parallelism()` counts hyperthreads too, which
/// overstates cores on any SMT-enabled host.
fn linux_cpuinfo() -> (Option<String>, Option<usize>) {
    let Ok(text) = fs::read_to_string("/proc/cpuinfo") else {
        return (None, None);
    };

    let mut model = None;
    let mut pairs: HashSet<(String, String)> = HashSet::new();
    let mut phys_id: Option<String> = None;
    let mut core_id: Option<String> = None;
    for line in text.lines() {
        if line.trim().is_empty() {
            if let (Some(p), Some(c)) = (phys_id.take(), core_id.take()) {
                pairs.insert((p, c));
            }
            continue;
        }
        let Some((key, value)) = line.split_once(':') else {
            continue;
        };
        let value = value.trim().to_string();
        match key.trim().to_lowercase().as_str() {
            "model name" if model.is_none() => model = Some(value),
            "physical id" => phys_id = Some(value),
            "core id" => core_id = Some(value),
            _ => {}
        }
    }
    if let (Some(p), Some(c)) = (phys_id, core_id) {
        pairs.insert((p, c));
    }
    (model, (!pairs.is_empty()).then_some(pairs.len()))
}

fn linux_mem_total_bytes() -> Option<u64> {
    let text = fs::read_to_string("/proc/meminfo").ok()?;
    for line in text.lines() {
        if let Some(rest) = line.strip_prefix("MemTotal:") {
            let kb: u64 = rest.split_whitespace().next()?.parse().ok()?;
            return Some(kb * 1024);
        }
    }
    None
}

fn os_description() -> String {
    if let Ok(text) = fs::read_to_string("/etc/os-release") {
        for line in text.lines() {
            if let Some(rest) = line.strip_prefix("PRETTY_NAME=") {
                return rest.trim().trim_matches('"').to_string();
            }
        }
    }
    std::env::consts::OS.to_string()
}

/// Queries `nvidia-smi` if it's on `PATH`. Absent binary, no GPU, or a
/// failed/non-zero call all just mean "no GPUs to report" — this must never
/// turn a CPU-only host into a request failure.
fn nvidia_gpus() -> Vec<GpuInfo> {
    let Ok(output) = Command::new("nvidia-smi")
        .args([
            "--query-gpu=name,memory.total,driver_version",
            "--format=csv,noheader,nounits",
        ])
        .output()
    else {
        return Vec::new();
    };
    if !output.status.success() {
        return Vec::new();
    }

    let text = String::from_utf8_lossy(&output.stdout);
    let mut gpus = Vec::new();
    for line in text.lines() {
        let parts: Vec<&str> = line.split(',').map(str::trim).collect();
        let [name, mem, ..] = parts.as_slice() else {
            continue;
        };
        if name.is_empty() {
            continue;
        }
        gpus.push(GpuInfo {
            name: name.to_string(),
            memory_total_mb: mem.parse::<f64>().ok().map(|v| v as u64),
            driver_version: parts
                .get(2)
                .filter(|v| !v.is_empty())
                .map(|v| v.to_string()),
        });
    }
    gpus
}

/// This process's host: the machine `armor-api` itself is running on.
pub fn local_hardware_info() -> HardwareInfo {
    let (model, physical_cores) = linux_cpuinfo();
    let logical_cores = std::thread::available_parallelism().ok().map(|n| n.get());
    let mem_total = linux_mem_total_bytes();

    HardwareInfo {
        cpu: CpuInfo {
            model,
            architecture: std::env::consts::ARCH.to_string(),
            logical_cores,
            physical_cores,
        },
        memory: MemoryInfo {
            total_bytes: mem_total,
        },
        gpus: nvidia_gpus(),
        os: os_description(),
    }
}
